test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -153,20 +189,29 @@ public TestCase upgradesFrom(Semver from)
         /** performs all supported upgrade paths that exist in between from and to (inclusive) **/
         public TestCase upgrades(Semver from, Semver to)
         {
-            SUPPORTED_UPGRADE_PATHS.stream()
-                .filter(upgradePath -> (upgradePath.left.compareTo(from) >= 0 && upgradePath.right.compareTo(to) <= 0))
-                .forEachOrdered(upgradePath ->
-                {
-                    this.upgrade.add(
-                            new TestVersions(versions.getLatest(upgradePath.left), versions.getLatest(upgradePath.right)));
-                });
+            List<TestVersions> upgrades = findUpgradePaths(from, to);
+            // order by most direct to largest path
+            Collections.sort(upgrades, Comparator.comparingInt(a -> a.upgrade.size()));
+            logger.info("Adding upgrades of\n{}", upgrades.stream().map(TestVersions::toString).collect(Collectors.joining("\n")));
+            this.upgrade.addAll(upgrades);
             return this;
         }
 
+        private List<TestVersions> findUpgradePaths(Semver from, Semver to)
+        {
+            return SUPPORTED_UPGRADE_PATHS.findPaths(from, to).stream()
+                                          .map(m -> new TestVersions(versions.getLatest(m.get(0)), m.stream().skip(1).map(versions::getLatest).collect(Collectors.toList())))
+                                          .collect(Collectors.toList());
+        }
+
         /** Will test this specific upgrade path **/
         public TestCase singleUpgrade(Semver from)
         {
-            this.upgrade.add(new TestVersions(versions.getLatest(from), versions.getLatest(CURRENT)));
+            TestVersions target = new TestVersions(versions.getLatest(from), Arrays.asList(versions.getLatest(CURRENT)));
+            Set<TestVersions> supported = new HashSet<>(findUpgradePaths(from, CURRENT));
+            if (!supported.contains(target))
```

Comment:
the thought process is that there may be cases in the future where from can't directly go to to, such as 2.1 -> 4.2.... this just makes sure we detect that when the test is running